File: mobile/lib/domain/entities/project.dart
class ProjectEntity {
  const ProjectEntity({
    required this.id,
    required this.name,
    required this.path,
    required this.env,
    required this.port,
    required this.entryCmd,
    required this.source,
    required this.status,
    this.gitRepo,
    this.gitBranch,
    this.group,
    this.favorite = false,
    this.lastOpened,
    this.createdAt,
    this.runningPort,
    this.tunnelUrl,
  });

  final String id;
  final String name;
  final String path;
  final String env;
  final int port;
  final String entryCmd;
  final String source;
  final String status;
  final String? gitRepo;
  final String? gitBranch;
  final String? group;
  final bool favorite;
  final DateTime? lastOpened;
  final DateTime? createdAt;
  final int? runningPort;
  final String? tunnelUrl;

  factory ProjectEntity.fromJson(Map<String, dynamic> json) {
    DateTime? ts(String key) {
      final raw = json[key];
      if (raw == null) return null;
      if (raw is String && raw.isNotEmpty) return DateTime.tryParse(raw);
      return null;
    }

    return ProjectEntity(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      path: json['path'] as String? ?? '',
      env: json['env'] as String? ?? 'generic',
      port: (json['port'] as num?)?.toInt() ?? 0,
      entryCmd: json['entry_cmd'] as String? ?? '',
      source: json['source'] as String? ?? 'empty',
      status: json['status'] as String? ?? 'stopped',
      gitRepo: json['git_repo'] as String?,
      gitBranch: json['git_branch'] as String?,
      group: json['group'] as String?,
      favorite: json['favorite'] as bool? ?? false,
      lastOpened: ts('last_opened'),
      createdAt: ts('created_at'),
      runningPort: (json['running_port'] as num?)?.toInt(),
      tunnelUrl: json['tunnel_url'] as String?,
    );
  }

  bool get isRunning => status == 'running';
  bool get isStarting => status == 'starting';
  bool get hasError => status == 'error';
}

File: backend/internal/domain/project/project.go
// Package project owns the Project entity + lifecycle orchestration.
// Runtime fields (Status, RunningPort, TunnelURL) are not persisted — they
// are assembled from the runner + tunnel services at read time.
package project

import "time"

// Project matches the JSON shape in plan.md §4.3.
type Project struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Path        string    `json:"path"`
	Env         string    `json:"env"`
	Port        int       `json:"port"`
	EntryCmd    string    `json:"entry_cmd"`
	Source      string    `json:"source"`
	GitRepo     string    `json:"git_repo,omitempty"`
	GitBranch   string    `json:"git_branch,omitempty"`
	Group       string    `json:"group,omitempty"`
	Favorite    bool      `json:"favorite"`
	LastOpened  time.Time `json:"last_opened"`
	CreatedAt   time.Time `json:"created_at"`
	Status      string    `json:"status"`
	RunningPort int       `json:"running_port,omitempty"`
	TunnelURL   string    `json:"tunnel_url,omitempty"`
}

// Valid project statuses. The DB never stores these — they are runtime-only.
const (
	StatusStopped  = "stopped"
	StatusStarting = "starting"
	StatusRunning  = "running"
	StatusError    = "error"
)
